### PR TITLE
Minimal clone implementation in ThreadPtrace

### PR DESCRIPTION
**Pull request by sporksmith**
* Add clone tests
* Refactor Process to handle multiple threads
* Implement clone handler for ThreadPtrace

### CODE REVIEW

**Comment by sporksmith** [comment]
Progress on #826

File: src/main/host/syscall/clone.h [hunk 1]
```diff
@@ -0,0 +1,9 @@
+/*
+ * The Shadow Simulator
+ * See LICENSE for licensing information
+ */
+
+#include "main/host/syscall_handler.h"
+#include "main/host/syscall_types.h"
+
+SysCallReturn syscallhandler_clone(SysCallHandler* sys, const SysCallArgs* args);
```
Comment (robgjansen) [inline src/main/host/syscall/clone.h:9]: Do you want to be consistent with the other handlers and add an include guard and use the `SYSCALL_HANDLER` define? E.g.:

```c
#ifndef SRC_MAIN_HOST_SYSCALL_CLONE_H_
#define SRC_MAIN_HOST_SYSCALL_CLONE_H_

#include "main/host/syscall/protected.h"

SYSCALL_HANDLER(clone);

#endif /* SRC_MAIN_HOST_SYSCALL_CLONE_H_ */
```

File: src/main/host/thread.h [hunk 2]
```diff
@@ -94,4 +94,12 @@ uint32_t thread_getHostId(Thread* thread);
 pid_t thread_getNativePid(Thread* thread);
 pid_t thread_getNativeTid(Thread* thread);
 
+int thread_getID(Thread* thread);
+int thread_clone(Thread* thread, unsigned long flags, PluginPtr child_stack, PluginPtr ptid,
+                 PluginPtr ctid, unsigned long newtls, Thread** child);
+
+void thread_setTidAddress(Thread* thread, PluginPtr addr);
+
+bool thread_isLeader(Thread* thread);
```
Comment (robgjansen) [inline src/main/host/thread.h:103]: Do we want docs for these?

File: src/test/CMakeLists.txt [hunk 3]
```diff
@@ -63,8 +64,9 @@ add_test(
...
-set_tests_properties(shadow-leakcheck-grep PROPERTIES DEPENDS "bind-shadow-preload;bind-shadow-ptrace;cpp-shadow-preload;cpp-shadow-ptrace;epoll-shadow-ptrace;epoll-writeable-shadow-ptrace;file-shadow-ptrace;mmap-shadow-ptrace;phold-shadow-ptrace;shutdown-shadow-preload;shutdown-shadow-ptrace;sleep-shadow-preload;sleep-shadow-ptrace;udp-shadow-ptrace;udp-uniprocess-shadow-preload;udp-uniprocess-shadow-ptrace;unistd-shadow-preload;unistd-shadow-ptrace")
+# All shadow test targets should be listed as dependencies, so that the leakcheck can grep their output for leaks.
+# Generate with: `(cd build ; ctest -R shadow -E leakcheck -N | head --lines=-2 | tail --lines=+2 | awk '{printf $3 ";"}')`
```
Comment (robgjansen) [inline src/test/CMakeLists.txt:68]: OMG thank you! 🥇

File: src/test/clone/test_clone.c [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    return 0;
+}
+
+static void _clone_child_exits_after_leader() {
+    // allocate some memory for the cloned thread.
+    uint8_t* stack = calloc(CLONE_TEST_STACK_NBYTES, 1);
+
+    // Heap grows up,
+    // Stack grows down,
+    // Head, shoulders, knees, and toes...
```
Comment (robgjansen) [inline src/test/clone/test_clone.c:88]: Compress and combine to one line?

File: src/test/clone/test_clone.c [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+int main(int argc, char** argv) {
+    g_test_init(&argc, &argv, NULL);
+
+    g_test_add("/clone/clone_minimal", void, NULL, NULL, _clone_minimal, NULL);
+    // FIXME: currently hangs under shadow. The exit_group sycall is executed
+    // natively from the thread leader, which does cause all threads to exit.
+    // In Shadow, however, the corresponding Shadow thread remains blocked by
+    // its nanosleep syscall. We probably want to add a handler for the
+    // exit_group syscall to wake-up/terminate the other threads in the
+    // process.
```
Comment (robgjansen) [inline src/test/clone/test_clone.c:119]: +1

File: src/main/host/syscall/clone.c [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    if (flags & CLONE_CHILD_CLEARTID) {
+        flags &= ~CLONE_CHILD_CLEARTID;
+        thread_setTidAddress(child, ctid);
+    }
+
+    if (flags) {
+        warning("Unhandled clone flags %ld", flags);
+    }
+
+    process_addThread(sys->process, child);
```
Comment (robgjansen) [inline src/main/host/syscall/clone.c:67]: Just wanted to double check my understanding: the above `thread_clone` call will cause us to create the native thread in the plugin, but we won't tell that thread to run yet until we call `process_addThread` and later `thread_resume`. I think that's the correct sequence to ensure that the process class knows about the thread we are creating before that thread starts calling syscalls.

**Review by robgjansen — APPROVED** [review]
LGTM modulo some minor nits.

File: src/main/host/thread.h (hunk 2 above)
Comment (sporksmith) [reply src/main/host/thread.h]: Done

File: src/test/clone/test_clone.c (hunk 4 above)
Comment (sporksmith) [reply src/test/clone/test_clone.c]: done.

**Pull request merged**